### PR TITLE
ci: change build-and-test workflow

### PR DESCRIPTION
**Pull request by KeisukeShima**
Signed-off-by: Keisuke Shima <19993104+KeisukeShima@users.noreply.github.com>

## Related Issue(required)

https://github.com/autowarefoundation/autoware.universe/pull/292#discussion_r791302632

## Description(required)

There are two reasons:
* Detect any side effects caused by the added/removed code from PRs.
* Calculate overall coverage for CodeCov.

## Review Procedure(required)

Make sure that the build-and-test workflow works on `tier4/proposal` branch after this PR has been merged.

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [commit-guidelines][commit-guidelines]
- [x] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: .github/workflows/build-and-test.yaml [hunk 1]
```diff
@@ -1,6 +1,9 @@
 name: build-and-test-scheduled
 
 on:
+  push:
+    branches:
+      - tier4/main
```
Comment (kenji-miyake) [inline .github/workflows/build-and-test.yaml:6]: There is no branch named `tier4/main`.
We only have `tier4/universe`(temporary) and `main` here.

**Comment by KeisukeShima** [comment]
After this PR is merged, I will add codecov-action step to build-and-test workflow on #292

File: .github/workflows/build-and-test.yaml (hunk 1 above)
Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:6]: Oh, thanks.

Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml]: Fixed branch name. 9deab047

File: .github/workflows/build-and-test.yaml [hunk 2]
```diff
@@ -1,6 +1,10 @@
 name: build-and-test-scheduled
 
 on:
+  push:
+    branches:
+      - main
+      - tier4/proposal
```
Comment (kenji-miyake) [inline .github/workflows/build-and-test.yaml:7]: Please tell me about the database of CoveCov.
Are the results stored for branches individually?

File: .github/workflows/build-and-test.yaml [hunk 3]
```diff
@@ -1,6 +1,10 @@
 name: build-and-test-scheduled
 
 on:
+  push:
```
Comment (kenji-miyake) [inline .github/workflows/build-and-test.yaml:4]: `build-and-test-scheduled` takes about 1~2 hours.
What happens if new PRs are submitted before `build-and-test-scheduled` completes?

File: .github/workflows/build-and-test.yaml (hunk 2 above)
Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:7]: Yes., according to the documentation,
* Branches will show up when a commit is uploaded on that branch.
* Pull requests are synced, regardless of having a coverage report uploaded. The sync will occur upon the first pull request webhook sent from GitHub to Codecov.
* Codecov only shows commits that have uploaded coverage reports and are six months or less old.

https://docs.codecov.com/docs/frequently-asked-questions#where-are-all-my-commitsbranchespulls
https://docs.codecov.com/docs/frequently-asked-questions#where-are-my-older-commits-my-project-dashboard-doesnt-show-any-commit-data-in-the-line-graph

Comment (kenji-miyake) [reply .github/workflows/build-and-test.yaml:7]: Thank you.
What do you mean by `Pull requests are synced`?

File: .github/workflows/build-and-test.yaml (hunk 3 above)
Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:4]: In this case, CodeCov cannot find base commit. 
CodeCov will display the following warnings in its comment: https://github.com/autowarefoundation/autoware.universe/pull/292#issuecomment-1020797172

File: .github/workflows/build-and-test.yaml (hunk 2 above)
Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:7]: Maybe CodeCov is synchronizing pull requests in order to be able to use pull request related features:

https://docs.codecov.com/docs/pull-request-comments
https://docs.codecov.com/docs/commit-status#excluding-tests-example
https://docs.codecov.com/docs/github-checks

File: .github/workflows/build-and-test.yaml [hunk 4]
```diff
@@ -1,12 +1,16 @@
-name: build-and-test-scheduled
+name: build-and-test
```
Comment (kenji-miyake) [inline .github/workflows/build-and-test.yaml:1]: Why did you change here?

File: .github/workflows/build-and-test.yaml [hunk 5]
```diff
@@ -1,12 +1,16 @@
...
+    branches:
+      - main
+      - tier4/proposal
   schedule:
     - cron: 0 19 * * * # run at 4 AM JST
   workflow_dispatch:
 
 jobs:
-  build-and-test-scheduled:
+  build-and-test:
```
Comment (kenji-miyake) [inline .github/workflows/build-and-test.yaml:13]: Why did you change here?

File: .github/workflows/build-and-test.yaml (hunk 4 above)
Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:1]: This is because I added push events, this workflow is no longer just for scheduled jobs.

File: .github/workflows/build-and-test.yaml (hunk 5 above)
Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:13]: The reason is the same as follwing thread. Let's discuss here:
https://github.com/autowarefoundation/autoware.universe/pull/297/files#r792369262

File: .github/workflows/build-and-test.yaml (hunk 4 above)
Comment (kenji-miyake) [reply .github/workflows/build-and-test.yaml:1]: I understand.
I'm not sure whether to add the push event trigger yet, so let's discuss it in this thread.
https://github.com/autowarefoundation/autoware.universe/pull/297#discussion_r791468797

File: .github/workflows/build-and-test.yaml (hunk 3 above)
Comment (kenji-miyake) [reply .github/workflows/build-and-test.yaml:4]: Then, I feel there is no reason for adding push events. :thinking: 
Do we really need it? Could you explain the reason again, please?

File: .github/workflows/build-and-test.yaml (hunk 2 above)
Comment (kenji-miyake) [reply .github/workflows/build-and-test.yaml:7]: Mm, what's `synchronizing`?

File: .github/workflows/build-and-test.yaml (hunk 3 above)
Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:4]: You are probably concerned that adding a push event will not get the correct Diff.
In my opinion, only `build-and-test-scheduled` leaves incomplete coverage records, and CodeCov cannot create Diffs. That's why we need the push event.

Also, I found a new document that says:

Codecov checks all your CI statuses on **every commit** to make sure the coverage results are accurate. This process is very important to the inner working of Codecov and data accuracy.

https://docs.codecov.com/docs/ci-service-relationship#if-ci-is-pending-then-the-coverage-results

Therefore, we really need it.

Comment (kenji-miyake) [reply .github/workflows/build-and-test.yaml:4]: I see, but since we're using self-hosted runner for this, I don't want to run this workflow for every commit.
How about preparing a new workflow such as `update-coverage.yaml`?

Comment (kenji-miyake) [reply .github/workflows/build-and-test.yaml:4]: Or splitting the ARM build job from `build-and-test-scheduled.yaml` is okay.

Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:4]: Is running the coverage job on x64 machine acceptable to you?
I don't think CPU architecture matters much for coverage, so I think it is enough to run it on x64 machine.

Comment (kenji-miyake) [reply .github/workflows/build-and-test.yaml:4]: Yes, it's fine.

Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:4]: Split a workflow file.
b5a1da31

**Review by kenji-miyake — DISMISSED** [review]

File: .github/workflows/build-and-test.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  schedule:
+    - cron: 0 19 * * * # run at 4 AM JST
+  workflow_dispatch:
+
+jobs:
+  build-and-test:
+    runs-on: ubuntu-latest
+    container: ros:galactic
+    steps:
+      - name: Checkout repository
```
Comment (kenji-miyake) [inline .github/workflows/build-and-test.yaml:17]: https://github.com/autowarefoundation/autoware.universe/pull/307

```suggestion
      - name: Check out repository
```

File: .github/workflows/build-and-test.yaml (hunk 2 above)
Comment (KeisukeShima) [reply .github/workflows/build-and-test.yaml:7]: > Mm, what's `synchronizing`?

The documentation doesn't have more information about the word "synced", I assume CodeCov needs to track pull requests in order to provide the Pulls tab in the dashboard.

ex.
https://app.codecov.io/gh/KeisukeShima/autoware.universe/pulls?page=1&state=open&order=-pullid

**Review by kenji-miyake — APPROVED** [review]

File: .github/workflows/build-and-test.yaml (hunk 2 above)
Comment (kenji-miyake) [reply .github/workflows/build-and-test.yaml:7]: Hmm, it's difficult... But thank you for your explanations! :pray:
